### PR TITLE
Properly exclude packages from the install set

### PR DESCRIPTION
**Pull request by Conan-Kudo**
The current approach does not account for them being pulled in as rich or weak dependencies. By excluding them at the sack level, it should properly prevent packages from being installed when they shouldn't.

This approach was pulled from rhinstaller/livecd-tools@617e737474fb17da0e266422367e56ec54516bec and adapted for Anaconda.

### CODE REVIEW

**Comment by jkonecny12** [comment]
Jenkins, it's ok to test.

**Comment by jkonecny12** [comment]
jenkins, test this please

**Comment by jkonecny12** [comment]
Please ignore the tests. It's failing because there is a problem in repositories.

**Comment by jkonecny12** [comment]
jenkins, test this please

**Comment by jkonecny12** [comment]
jenkins, test this please

**Comment by ignatenkobrain** [comment]
https://github.com/rpm-software-management/dnf/commit/7b3498cc0e36918186b62967c0eda077fa04babd

Although for this thing you will need to bump requirement of DNF to 2.1

**Comment by rvykydal** [comment]
As @vpodzime guesses correctly we are still excluding packages in _select_group(). Not sure if it is a big problem but it can be missed when modifying the code in the future.

**Comment by jkonecny12** [comment]
Possible fix for bug https://bugzilla.redhat.com/show_bug.cgi?id=1412398 .

**Comment by hroncok** [comment]
Bump please?

File: pyanaconda/packaging/dnfpayload.py [hunk 1]
```diff
@@ -401,6 +401,10 @@ def _apply_selections(self):
             except packaging.NoSuchGroup as e:
                 self._miss(e)
 
+        for pkg_name in set(self.data.packages.excludedList):
```
[the remark this answers is not in the document]
Comment (Conan-Kudo) [reply pyanaconda/packaging/dnfpayload.py]: That is the original list...?

File: pyanaconda/packaging/dnfpayload.py [hunk 2]
```diff
@@ -401,6 +401,10 @@ def _apply_selections(self):
...
+            self._exclude_package(pkg_name)
+            logging.info("excluding package: '%s'", pkg_name)
+
         for pkg_name in set(self.data.packages.packageList) - set(self.data.packages.excludedList):
```
[the remark this answers is not in the document]
Comment (Conan-Kudo) [reply pyanaconda/packaging/dnfpayload.py]: Group exclusion differs from package exclusion, and cannot be excluded at the sack level.

File: pyanaconda/packaging/dnfpayload.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/packaging/dnfpayload.py]: Just iterate over `self.data.packages.excludedList`, please.

[the remark this answers is not in the document]
Comment (Conan-Kudo) [reply pyanaconda/packaging/dnfpayload.py]: I am iterating over it... This is a foreach loop. I'm also logging what's being excluded...

[the remark this answers is not in the document]
Comment (Conan-Kudo) [reply pyanaconda/packaging/dnfpayload.py]: Oh wait, you mean don't make it into a set?

File: pyanaconda/packaging/dnfpayload.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/packaging/dnfpayload.py]: The `exclude` parameter of the `group_install` method excludes packages from an installed group. It is not used to exclude groups: 

http://dnf.readthedocs.io/en/latest/api_base.html?highlight=group_in#dnf.Base.group_install

I have tested it and packages can be excluded from groups with `_exclude_package` as well.

**Comment by InStallker** [comment]
Can one of the admins verify this patch?

**Comment by Conan-Kudo** [comment]
I'm going to come back to this asap, I've been busy for a while and this got lost in the shuffle.

**Comment by InStallker** [comment]
Can one of the admins verify this patch?

**Comment by Conan-Kudo** [comment]
@poncovka I've rebased against current master and done the suggested changes.

**Comment by jkonecny12** [comment]
jenkins, test this please

**Comment by jkonecny12** [comment]
Please fix this:
```
************* Module ../pyanaconda/payload/dnfpayload.py
E0602(undefined-variable):455,12: DNFPayload._apply_selections: Undefined variable 'logging'
FAIL pylint/runpylint.py (exit status: 2)
```

**Comment by Conan-Kudo** [comment]
@jkonecny12 Should be fixed now.

**Comment by poncovka** [comment]
jenkins, test this please

**Pull request merged**